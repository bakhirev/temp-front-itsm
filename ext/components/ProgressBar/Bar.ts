import styled from 'styled-components';

import { DEFAULT_THEME } from '../common';

export const BORDER_RADIUS = 2;
const HEIGHT = 4;

export const Bar = styled.div`
  background: ${({ theme }) => theme.color.neutral[10]};
  border-radius: ${BORDER_RADIUS}px;
  height: ${HEIGHT}px;
  overflow: hidden;
`;

Bar.defaultProps = {
  theme: DEFAULT_THEME,
};
