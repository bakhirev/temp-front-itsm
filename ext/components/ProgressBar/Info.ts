import styled from 'styled-components';

import { DEFAULT_THEME } from '../common';

const MARGIN_BOTTOM = 8;

export const Info = styled.div<{ error: boolean }>`
  color: ${({ error, theme }) => (error ? theme.color.error[60] : theme.color.neutral[90])};
  display: flex;
  justify-content: space-between;
  margin-bottom: ${MARGIN_BOTTOM}px;
`;

Info.defaultProps = {
  theme: DEFAULT_THEME,
};
