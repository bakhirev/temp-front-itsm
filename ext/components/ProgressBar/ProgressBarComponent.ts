import styled from 'styled-components';

const HEIGHT = {
  page: 28,
  top: 4,
};

const MIN_WIDTH = 140;

export type Kind = 'page' | 'top';

export const ProgressBarComponent = styled.div<{ kind: Kind }>`
  height: ${({ kind }) => HEIGHT[kind]}px;
  min-width: ${MIN_WIDTH}px;
  width: 100%;
`;
