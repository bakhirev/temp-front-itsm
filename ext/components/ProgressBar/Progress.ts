import styled from 'styled-components';

import { DEFAULT_THEME } from '../common';

import { BORDER_RADIUS } from './Bar';

export const Progress = styled.div<{ error: boolean; value: number }>`
  background: ${({ error, theme }) => (error ? theme.color.error[60] : theme.color.primary[60])};
  border-radius: ${BORDER_RADIUS}px;
  height: 100%;
  width: ${({ value }) => value}%;
`;

Progress.defaultProps = {
  theme: DEFAULT_THEME,
};
