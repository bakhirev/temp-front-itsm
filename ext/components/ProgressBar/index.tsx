import React from 'react';
import type { FC } from 'react';

import { Caption1 } from '../Typography';

import { Bar } from './Bar';
import { Info } from './Info';
import { Progress } from './Progress';
import { ProgressBarComponent } from './ProgressBarComponent';
import type { Kind } from './ProgressBarComponent';

export interface IProgressBarProps {
  /** Имя класса для переопределения стилей */
  className?: string;
  /** data-test-id атрибут для тестирования компонента */
  dataTestId?: string;
  /** Флаг ошибки */
  error?: boolean;
  /** Вид компонента */
  kind?: Kind;
  /** Надписть над полосой прогресса */
  label?: string;
  /** Значение прогресса от 1 до 100 */
  value: number;
}

export const ProgressBar: FC<IProgressBarProps> = ({
  className,
  dataTestId,
  error = false,
  kind = 'page',
  label,
  value,
}) => (
  <ProgressBarComponent className={className} data-test-id={dataTestId} kind={kind}>
    {kind === 'page' && (
      <Info error={error}>
        <Caption1>{label}</Caption1>
        <Caption1>{`${value}%`}</Caption1>
      </Info>
    )}
    <Bar>
      <Progress error={error} value={value} />
    </Bar>
  </ProgressBarComponent>
);
