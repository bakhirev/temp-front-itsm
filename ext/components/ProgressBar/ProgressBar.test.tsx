import React from 'react';
import { shallow } from 'enzyme';

import { ProgressBar } from '../ProgressBar';

jest.mock('../common', () => ({
  DEFAULT_THEME: {},
}));

describe('ProgressBar', () => {
  it('should render component with kind "page" by default', () => {
    const wrapper = shallow(<ProgressBar value={0} />);
    expect(wrapper).toMatchSnapshot();
  });

  it('should render component with kind "top"', () => {
    const wrapper = shallow(<ProgressBar kind="top" value={0} />);
    expect(wrapper).toMatchSnapshot();
  });

  it('should render component with label when kind equals "page"', () => {
    const wrapper = shallow(<ProgressBar kind="page" label="Загрузка данных ..." value={0} />);
    expect(wrapper).toMatchSnapshot();
  });

  it('should render component with percentage progress when kind equals "page"', () => {
    const wrapper = shallow(<ProgressBar kind="page" value={50} />);
    expect(wrapper).toMatchSnapshot();
  });

  it('should render component with error when kind equals "page"', () => {
    const wrapper = shallow(<ProgressBar error kind="page" value={0} />);
    expect(wrapper).toMatchSnapshot();
  });
});
