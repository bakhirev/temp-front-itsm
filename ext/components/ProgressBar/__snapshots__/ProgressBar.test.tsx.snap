// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ProgressBar should render component with error when kind equals "page" 1`] = `
<styled.div
  kind="page"
>
  <styled.div
    error={true}
    theme={Object {}}
  >
    <styled.div />
    <styled.div>
      0%
    </styled.div>
  </styled.div>
  <styled.div
    theme={Object {}}
  >
    <styled.div
      error={true}
      theme={Object {}}
      value={0}
    />
  </styled.div>
</styled.div>
`;

exports[`ProgressBar should render component with kind "page" by default 1`] = `
<styled.div
  kind="page"
>
  <styled.div
    error={false}
    theme={Object {}}
  >
    <styled.div />
    <styled.div>
      0%
    </styled.div>
  </styled.div>
  <styled.div
    theme={Object {}}
  >
    <styled.div
      error={false}
      theme={Object {}}
      value={0}
    />
  </styled.div>
</styled.div>
`;

exports[`ProgressBar should render component with kind "top" 1`] = `
<styled.div
  kind="top"
>
  <styled.div
    theme={Object {}}
  >
    <styled.div
      error={false}
      theme={Object {}}
      value={0}
    />
  </styled.div>
</styled.div>
`;

exports[`ProgressBar should render component with label when kind equals "page" 1`] = `
<styled.div
  kind="page"
>
  <styled.div
    error={false}
    theme={Object {}}
  >
    <styled.div>
      Загрузка данных ...
    </styled.div>
    <styled.div>
      0%
    </styled.div>
  </styled.div>
  <styled.div
    theme={Object {}}
  >
    <styled.div
      error={false}
      theme={Object {}}
      value={0}
    />
  </styled.div>
</styled.div>
`;

exports[`ProgressBar should render component with percentage progress when kind equals "page" 1`] = `
<styled.div
  kind="page"
>
  <styled.div
    error={false}
    theme={Object {}}
  >
    <styled.div />
    <styled.div>
      50%
    </styled.div>
  </styled.div>
  <styled.div
    theme={Object {}}
  >
    <styled.div
      error={false}
      theme={Object {}}
      value={50}
    />
  </styled.div>
</styled.div>
`;
